devtools/x/src/lint/workspace_classify.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::config::TestOnlyConfig;
+use guppy::graph::{BuildTargetId, BuildTargetKind};
+use indoc::indoc;
+use x_lint::prelude::*;
+
+/// Ensure that every package in the workspace is classified as either a default member or test-only.
+#[derive(Debug)]
+pub struct DefaultOrTestOnly<'cfg> {
+    config: &'cfg TestOnlyConfig,
+}
+
+impl<'cfg> DefaultOrTestOnly<'cfg> {
+    pub fn new(config: &'cfg TestOnlyConfig) -> Self {
+        Self { config }
+    }
+}
+
+impl<'cfg> Linter for DefaultOrTestOnly<'cfg> {
+    fn name(&self) -> &'static str {
+        "default-or-test-only"
+    }
+}
+
+impl<'cfg> PackageLinter for DefaultOrTestOnly<'cfg> {
+    fn run<'l>(
+        &self,
+        ctx: &PackageContext<'l>,
+        out: &mut LintFormatter<'l, '_>,
+    ) -> Result<RunStatus<'l>> {
+        let default_members = ctx.project_ctx().default_members()?;
+        let package = ctx.metadata();
+
+        if self.config.members.contains(ctx.workspace_path()) {
+            // This package is explicitly marked as test-only.
+            return Ok(RunStatus::Executed);
+        }
+
+        let binary_kind = package
+            .build_targets()
+            .filter_map(|target| {
+                if matches!(target.id(), BuildTargetId::Binary(_)) {
+                    return Some("binary");
+                }
+                // If this is the library target, then look for the first binary-equivalent-kind
+                if let (BuildTargetId::Library, BuildTargetKind::LibraryOrExample(crate_types)) =
+                    (target.id(), target.kind())
+                {
+                    return crate_types
+                        .iter()
+                        .filter_map(|crate_type| {
+                            // These library types are equivalent to binaries.
+                            if crate_type == "cdylib"
+                                || crate_type == "dylib"
+                                || crate_type == "staticlib"
+                            {
+                                Some(crate_type.as_str())
+                            } else {
+                                None
+                            }
+                        })
+                        .next();
+                }
+                return None;
+            })
+            .next();
+
+        let default_root = default_members.is_root_member(package.id());
+        let default_reachable = default_members.contains(package.id());
+
+        match (binary_kind, default_root, default_reachable) {
```

Comment:
Here's the core decision table that determines where things should go.

Reply:
good pruning mechanism.  dig it.